File: value/src/main/java/com/bnorm/barkeep/model/RecipeValue.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model;

import javax.annotation.Nullable;

import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;

@AutoValue
public abstract class RecipeValue implements Recipe {

  public static Builder builder() {
    return new AutoValue_RecipeValue.Builder();
  }

  public static JsonAdapter<RecipeValue> jsonAdapter(Moshi moshi) { return new AutoValue_RecipeValue.MoshiJsonAdapter(moshi); }

  @Override
  @Nullable
  public abstract ImmutableSet<Component> getComponents();

  @AutoValue.Builder
  public abstract static class Builder {

    public abstract RecipeValue build();

    public abstract Builder setId(Long id);

    public abstract Builder setTitle(String title);

    public abstract Builder setDescription(String description);

    public abstract Builder setOwner(User owner);

    public abstract Builder setImageUrl(String imageUrl);

    public abstract Builder setInstructions(String instructions);

    public abstract Builder setSource(String source);

    abstract ImmutableSet.Builder<Component> componentsBuilder();

    public Builder addComponent(Component component) {
      componentsBuilder().add(component);
      return this;
    }

    public ComponentValueBuilder newComponentBuilder() {
      return new ComponentValueBuilder(this);
    }
  }


  public static class ComponentValueBuilder extends ComponentValue.Builder {

    private final RecipeValue.Builder parent;
    private final ComponentValue.Builder builder;

    public ComponentValueBuilder(Builder parent) {
      this.parent = parent;
      this.builder = ComponentValue.builder();
    }

    public Builder finish() {
      parent.addComponent(build());
      return parent;
    }

    @Override
    ComponentValue build() {
      return builder.build();
    }

    @Override
    public ComponentValueBuilder setIngredient(Ingredient ingredient) {
      builder.setIngredient(ingredient);
      return this;
    }

    @Override
    public ComponentValueBuilder setMin(double min) {
      builder.setMin(min);
      return this;
    }

    @Override
    public ComponentValueBuilder setMax(Double max) {
      builder.setMax(max);
      return this;
    }

    @Override
    public ComponentValueBuilder setComponentNum(long componentNum) {
      builder.setComponentNum(componentNum);
      return this;
    }

    @Override
    public ComponentValueBuilder setOrder(long order) {
      builder.setOrder(order);
      return this;
    }
  }
}

File: model-value/src/main/kotlin/com/bnorm/barkeep/model/RecipeValue.kt
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.model

import com.squareup.moshi.FromJson
import com.squareup.moshi.ToJson

data class RecipeSpecValue(override val title: String?,
                           override val description: String? = null,
                           override val owner: User? = null,
                           override val imageUrl: String? = null,
                           override val instructions: String? = null,
                           override val source: String? = null,
                           override val components: Set<Component>? = emptySet()) : RecipeSpec

data class RecipeValue(override val id: Long = -1,
                       override val title: String?,
                       override val description: String? = null,
                       override val owner: User? = null,
                       override val imageUrl: String? = null,
                       override val instructions: String? = null,
                       override val source: String? = null,
                       override val components: Set<Component>? = emptySet()) : Recipe

object RecipeValueAdapter {
  @ToJson fun toJson(recipe: RecipeSpec): RecipeSpecValue {
    return RecipeSpecValue(recipe.title,
                           recipe.description,
                           recipe.owner,
                           recipe.imageUrl,
                           recipe.instructions,
                           recipe.source,
                           recipe.components)
  }

  @FromJson fun fromJson(json: RecipeSpecValue): RecipeSpec {
    return json
  }

  @ToJson fun toJson(recipe: Recipe): RecipeValue {
    return RecipeValue(recipe.id,
                       recipe.title,
                       recipe.description,
                       recipe.owner,
                       recipe.imageUrl,
                       recipe.instructions,
                       recipe.source,
                       recipe.components)
  }

  @FromJson fun fromJson(json: RecipeValue): Recipe {
    return json
  }
}
